optuna/study/study.py

```diff
@@ -45,6 +50,51 @@
 _logger = logging.get_logger(__name__)
 
 
+_T = TypeVar("_T")
+_P = ParamSpec("_P")
+
+
+def _convert_positional_args(
+    *,
+    previous_positional_arg_names: Sequence[str],
+) -> Callable[[Callable[_P, _T]], Callable[_P, _T]]:
+    def decorator(func: Callable[_P, _T]) -> Callable[_P, _T]:
+        assert (
+            previous_positional_arg_names
+            == list(signature(func).parameters)[: len(previous_positional_arg_names)]
+        ), (
+            f"{previous_positional_arg_names} is not a sublist of"
+            f" {list(signature(func).parameters)}"
+        )
+
+        @wraps(func)
+        def wrapper(*args: Any, **kwargs: Any) -> _T:
+            if len(args) >= 1:
+                warnings.warn(
+                    f"{func.__name__}(): Positional arguments are deprecated."
+                    " Please give all values as keyword arguments."
+                )
```

Comment:
Ah, we should use a `FutureWarning` here, similar to https://github.com/optuna/optuna/blob/release-v3.0.0-a1/optuna/_deprecated.py.
```suggestion
                warnings.warn(
                    f"{func.__name__}(): Positional arguments are deprecated."
                    " Please give all values as keyword arguments.",
                    FutureWarning
                )
```

Reply:
Fix in 81504f6 and introduce `stacklevel` to indicate proper position in stack trace in 7229d24. The `warning_stacklevel` is required when decorators are nested.